### PR TITLE
[ML/Train] `TensorflowTrainer` implementation

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Implements `TensorflowTrainer`. Depends on https://github.com/ray-project/ray/pull/23211 (review only files with `tensorflow` in the name).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+from ray.train.tensorflow import prepare_dataset_shard
+from ray.ml.train.integrations.tensorflow import TensorflowTrainer
+from ray.ml.predictors.integrations.tensorflow import TensorflowPredictor
+from ray.ml.constants import MODEL_KEY, TRAIN_DATASET_KEY
+from ray.ml.result import Result
+
+
+class TrainCheckpointReportCallback(Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        train.save_checkpoint(**{MODEL_KEY: self.model.get_weights()})
```
Comment (amogkam) [inline python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py:21]: for these examples, can just use "model" directly? I think we've decided not to expose these constants to users for now.

File: python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+
+def get_dataset_pipeline(a=5, b=10, size=1000) -> DatasetPipeline:
+    def get_dataset(a, b, size) -> Dataset:
+        items = [i / size for i in range(size)]
+        dataset = ray.data.from_items([{"x": x, "y": a * x + b} for x in items])
+        return dataset
+
+    dataset = get_dataset(a, b, size)
+    train_dataset_pipeline = dataset.repeat().random_shuffle_each_window()
```
Comment (amogkam) [inline python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py:32]: Trainer is not supposed to support DatasetPipelines for now. Can we change this to a regular Dataset (remove the repeat and random shuffle)?

File: python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    with strategy.scope():
+        # Model building/compiling need to be within `strategy.scope()`.
+        multi_worker_model = build_model()
+        multi_worker_model.compile(
+            optimizer=tf.keras.optimizers.SGD(learning_rate=config.get("lr", 1e-3)),
+            loss=tf.keras.losses.mean_squared_error,
+            metrics=[tf.keras.metrics.mean_squared_error],
+        )
+
+    dataset_pipeline = train.get_dataset_shard(TRAIN_DATASET_KEY)
```
Comment (amogkam) [inline python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py:62]: Same thing here- can we just change this to "train" instead of TRAIN_DATASET_KEY?

File: python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    print(results.metrics)
+    return results
+
+
+def predict_linear(result: Result) -> Dataset:
+    predictor = TensorflowPredictor.from_checkpoint(result.checkpoint, build_model)
+    items = [{"x": np.random.uniform(0, 1)}] * 10
+    prediction_dataset = ray.data.from_items(items)
+
+    predictions = prediction_dataset.map_batches(
```
Comment (amogkam) [inline python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py:105]: Can we modify how we do prediction to show the "recommended" approach:

```python
checkpoint_object_ref = result.checkpoint.to_object_ref()

class TFScorer:
    def __init__(self):
        self.predictor = TensorflowPredictor.from_checkpoint(Checkpoint.from_object_ref(checkpoint_object_ref))
 
    def __call__(self, batch):
        self.predictor.predict(batch, dtype=tf.float32)

predictions = prediction_dataset.map_batches(TFScorer, compute="actors")
```

File: python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+from ray.data import Dataset
+from ray.data.dataset_pipeline import DatasetPipeline
+from ray.train.tensorflow import prepare_dataset_shard
+from ray.ml.train.integrations.tensorflow import TensorflowTrainer
+from ray.ml.predictors.integrations.tensorflow import TensorflowPredictor
+from ray.ml.constants import MODEL_KEY, TRAIN_DATASET_KEY
+from ray.ml.result import Result
+
+
+class TrainCheckpointReportCallback(Callback):
```
Comment (amogkam) [inline python/ray/ml/examples/tensorflow/tensorflow_linear_dataset_example.py:19]: Should we just move this to ml/integrations/tensorflow?

**Review by amogkam — APPROVED** [review]
LGTM! Just left some minor comments. Once DataParallelTrainer is merged in, I will merge master into this branch.

**Review by richardliaw — APPROVED** [review]
